### ISSUE TITLE
[Question]:  two factor authentication handling

### ISSUE THREAD

**Issue by greggo74** [bropat/eufy-security-ws#137]
### Ask your question

How do I enter the 2 factor authentication code for my docker container?

This is from the logs:
2022-09-26 12:38:17.496  INFO  Requested verification code for 2FA 
2022-09-26 12:38:21.132  INFO  Client disconnected with ip: 172.17.0.1 port: 35814 code: 1000 reason:  

I get emails with the code but it's requesting codes too quickly and I cannot work out how to add them to the docker instance.

Please advise.

**Comment by greggo74** [comment]
I tried adding the TRUSTED_DEVICE_NAME to the docker environment parameter and using my phone ID as the trusted device from the app but I had no luck.

**Comment by Norwegiandroid** [comment]
I have the excact same issue. Anyone else have an idea how to solve this?

**Comment by greggo74** [comment]
My short term solution is to remove 2-factor authentication, and i'm rolling, but I don't wait to leave it that way.

**Comment by Norwegiandroid** [comment]
> My short term solution is to remove 2-factor authentication, and i'm rolling, but I don't wait to leave it that way.

Thanks. Worked for me as well. I set up another eufy account without 2fa and shared my home with that account. Used those credentials in my eufy docker container and magic happened 🙂

**Comment by bropat** [comment]
@greggo74 @Norwegiandroid 

do you only use eufy-security-ws? Or do you use it through any integration?

**Comment by greggo74** [comment]
I run eufy-security-ws as a docker on the same device running home assistant. 
eufy-security-ws runs fine when I turn off two-factor authentic.
If I enable two-factor in my account I stay having the issues with the verification code. 
My home assistant eufy integration it configured to use my eufy-security-ws instance.

**Comment by bropat** [comment]
2FA has been working here for a long time, if it doesn't work through Home Assistant integration, you need to open an issue there and not here.

**Comment by St3faan** [comment]
I use "myphone" as environment variable and it worked.

**Comment by StoneTrip** [comment]
Using client.ts you can query for your verification code like this
once the client.ts is up and running query the verification like this
driver.set_verify_cod XXXXXX
it will send the verification code to the server and validate your login. no need to disable the 2FA

**Comment by karlg100** [comment]
@StoneTrip - can you provide a bit more context?  2FA Documentation for this project is extremely lacking.

**Comment by gazm2k5** [comment]
Can someone please provide simple instructions on how to enter the 2FA code in the container?

**Comment by rotweiss4** [comment]
I need also help how to enter 2FA. Is anybody out there how know it?

**Comment by artkrzy** [comment]
Same here, how do I pass 2FA to the config?

**Comment by leotantantan** [comment]
> Once you configure add-on and you can see in the logs message to provide MFA code, start integration. After a minute you should see the notification in setting with Eufy security addon needs attention, click on it and pop-up window will appear to provide MFA code

I can see the relevant messages in the log, but there is no pop-up in 'settings'. I have tried a few times, but it still doesn't work for me.

**Comment by CarnsJalone** [comment]
Hi there, not sure why this issue was closed, don't see a community resolution available here in the comments. Am I missing something?

**Comment by fuatakgun** [comment]
There are two possible flows underneath;
1. You have the wrong username, password and the add-on tried to login using the wrong credentials and failed. Add-on receives a captcha request but this captcha is not relevant or useful until credentials are fixed. Sometimes you just need to change the credentials even though they are right or correct. Please do not use too complex characters:-)
2. You have the right credentials and login requires captcha to continue. Integration pop-up will be useful to login in this flow to continue.

So, please check add-on logs to differentiate between 1 and 2.

**Comment by vomov** [comment]
I'm also running the websocket docker container by itself (no HA), and am at a loss on where to put the 2FA. I'd expect to do an HTTP PUT or something, but can't find whereto.

**Comment by vomov** [comment]
So, what if you're not running the integration, on account of not running HA?

**Comment by Pliind** [comment]
> So, what if you're not running the integration, on account of not running HA?

If youre not running HA i cannot help you.

**Comment by vomov** [comment]
> If youre not running HA i cannot help you.
I'm not sure what to do with this response.

The documentation (both in this repo as in the linked documentation) does not mention HA as a hard requirement. As this repo delivers a docker container that exposes eufy over a websocket, one would expect to be able to communicate over websockets. 

I see two possibilities:
1. @Pliind , you do not know how to communicate with the container over websockets. This is fine; I don't either. Maybe someone else knows; that's what I and several other users are hoping for.
2. the container itself has an undocumented hard dependency on HA, which is not documented.

**Comment by fuatakgun** [comment]
Captcha event: https://github.com/fuatakgun/eufy_security/blob/7e03591a993c817f60a4fc2f1c2b6cf39892a8ef/custom_components/eufy_security/eufy_security_api/api_client.py#L318

MFA event: https://github.com/fuatakgun/eufy_security/blob/7e03591a993c817f60a4fc2f1c2b6cf39892a8ef/custom_components/eufy_security/eufy_security_api/api_client.py#L322

Send captcha function: https://github.com/fuatakgun/eufy_security/blob/7e03591a993c817f60a4fc2f1c2b6cf39892a8ef/custom_components/eufy_security/eufy_security_api/api_client.py#L166

Send MFA code function: https://github.com/fuatakgun/eufy_security/blob/7e03591a993c817f60a4fc2f1c2b6cf39892a8ef/custom_components/eufy_security/eufy_security_api/api_client.py#L169

**Comment by SoleInvictus** [comment]
I also ran into this issue, so I'll lay out how to handle it in plain speech for my fellow "variable illumination" bulbs.

So you installed eufy-security-ws and observed how it nags you for 2FA in the logs. Good on you for being secure! I'm sure you looked up how to handle it online. You initially found no answers, only many unanswered questions. Maybe a couple of "I fixed it, thanks guys!" with no details of...how it was fixed. Luckily, an unemployed man with nothing else to do figured it out and decided to post about it. Here's what worked:

1. Install the [EUFY Security integration](https://github.com/fuatakgun/eufy_security) found in HACS. Instructions [here](https://github.com/fuatakgun/eufy_security/tree/7e03591a993c817f60a4fc2f1c2b6cf39892a8ef?tab=readme-ov-file#1-installing-eufy-security-add-on), including HACS install instructions. Be sure to restart HA at the end so the new integration starts!
2.  Return to the eufy-security-ws addon and restart it. Why? You only had 5 minutes to enter that 2FA and it has been what, 30 minutes? 60? Time to get another one.
3. In HA, go to Settings>Devices & Services. You'll see an error prompt at the top from Eufy Security, asking for your 2FA code. *Your body releases a shock of adrenaline. Could this...could this be it?*
4. Click it. *The tension mounts.*
5. Enter your code. *No. Fucking. Way. It worked?!*
6. High five your cat/dog/frog/self, you did it!
